### ISSUE TITLE
make it easier to test from PR: rewrite modrev ?

### ISSUE THREAD

**Issue by teto** [nvim-neorocks/luarocks-tag-release#26]
I am trying to test the workflow in a PR for nvim-treesitter:  https://github.com/nvim-treesitter/nvim-treesitter/actions/runs/4298698664/jobs/7493113344
 

One issue is that in PRs, the rockspec version is ref_name by default : 

``` action.yml
  version:
    description: "The version of your package. By default, the git tag is used."
    required: true
    default: ${{ github.ref_name }}
```
which in my example becomes "4403/merge" and breaks luarocks upload: 
```
building '/nix/store/dgccn4r9cf6q1znhvnp8h3lfmr9xqj3f-luarocks-tag-release.drv'...
/nix/store/93c8pf56d23wmjml80c7f7ga9q07b856-lua-5.1.5/bin/lua: ...rocks-tag-release/scm-1/bin/luarocks-tag-release.lua:101: Could not create nvim-treesitter-4403/merge-1.rockspec.
Publishing an untagged release.
Using template: nvim-treesitter-luarocks.template
Generating Luarocks release 4403/merge for: nvim-treesitter version 4403/merge from ref e04e08ab8de82eee3fa41c5b1a22c28b6f56d4aa.

Generated rockspec:
========================================================================================
local git_ref = 'e04e08ab8de82eee3fa41c5b1a22c28b6f56d4aa'
local modrev = '4403/merge'
```

This comment explains how github.ref_name works: https://stackoverflow.com/a/59780579

I am not sure what a good solution is, maybe:
- if running for a tag, keep the current default ref_name
- if running in a PR, use "dev" / "test" / "scm<short_sha>" or "scm<PR_NUMBER>" ?

NB: note to myself, I had completely missed the line:
```
        env:
          LUAROCKS_API_KEY: ${{ secrets.LUAROCKS_API_KEY }}
```
in examples. Turns out secrets are not accessible in PRs from forks so I've tried replacing the secret with a variable:
	`LUAROCKS_API_KEY: ${{ vars.LUAROCKS_API_KEY }}`

**Comment by mrcjkb** [comment]
Maybe [this section](https://github.com/nvim-neorocks/luarocks-tag-release/blob/master/bin/luarocks-tag-release.lua#L59) of the script can be modified for this.

```lua
local is_tag = os.getenv('GITHUB_REF_TYPE') == 'tag'
local git_ref = assert(os.getenv('GITHUB_REF_NAME'), 'GITHUB_REF_NAME not set')
local archive_dir_suffix = modrev
if not is_tag then
  print('Publishing an untagged release.')
  git_ref = assert(os.getenv('GITHUB_SHA'), 'GITHUB_SHA not set')
  archive_dir_suffix = git_ref
end
```
We just have to make sure that we maintain support for rolling releases (like the workflow used by fzf-lua).

The [`GITHUB_EVENT_NAME` environment variable](https://docs.github.com/en/actions/learn-github-actions/variables) can be used to determine how the workflow was triggered (e.g. if it's [`'pull_request'`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request)).

I would also suggest we only do the local install if the event is `'pull_request'` (and skip the upload and install from luarocks steps).

This would...

* allow us to add different use cases to our CI by using actions/checkout to checkout nvim-treesitter, fzf-lua, etc.
* allow plugins to test the luarocks install without uploading on `pull_request`.

**Comment by teto** [comment]
> We just have to make sure that we maintain support for rolling releases (like the workflow used by fzf-lua).

What does this mean ? they regularly override the uploaded current version ?

> I would also suggest we only do the local install if the event is 'pull_request' (and skip the upload and install from luarocks steps).

Being able to test the upload in the PR is really interesting, I would prefer to keep that. We dont want to override the actual package on luarocks.org though so the uploaded package name should be <NAME>-prXXXX for instance. And probably it should not appear in the root manifest, so remains as a "scm" version. 
The problem is that when testing and submitting a PR, one will need to pass their own LUAROCKS_API_KEY and uploaded packages will stay forever on luarocks. Maybe we could find a mechanism with luarocks.org folks for temporary packages, test upload.

Also why I raised the issue of easing testing upload from PRs in the first place: 
the nvim-treesitter repo needs a review from someone else before merging which is annoying in this iterative testing scenario. On another repo I would just merge until it succeeds (bad I know xD) but can't really do that with that repo.

**Comment by mrcjkb** [comment]
> What does this mean ? they regularly override the uploaded current version ?

It's a scheduled workflow (every night) that checks if there have been commits to `main` in the last 24 hours. If so, a package is published with version `0.0.<n_commits>`.

> Maybe we could find a mechanism with luarocks.org folks for temporary packages, test upload.

That sounds like a good idea.

> the nvim-treesitter repo needs a review from someone else before merging which is annoying in this iterative testing scenario. 

As a workaround for now, you can add a luarocks API key for a test account to your fork and push tags to your branch.

**Comment by mrcjkb** [comment]
Fixed by #63
